File: core/validator.py
import os
import subprocess

from core.detector import get_chdman_path


def verify_chd(filepath: str, on_log=None) -> bool:
    """Run chdman verify on a CHD file. Returns True if valid."""
    log = on_log or print
    if not os.path.isfile(filepath):
        log(f"[VERIFY] File not found: {filepath}")
        return False

    chdman = get_chdman_path()
    cmd = [chdman, "verify", "-i", filepath]
    log(f"[VERIFY] Checking {os.path.basename(filepath)}...")

    try:
        result = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            encoding="utf-8",
            errors="replace",
        )
        output = result.stdout.strip()
        for line in output.splitlines():
            log(f"  {line}")

        if result.returncode == 0:
            log("  ✅ CHD is valid")
            return True
        else:
            log("  ❌ CHD verification failed")
            return False
    except FileNotFoundError:
        log(f"  ❌ chdman not found at: {chdman}")
        return False


def get_chd_info(filepath: str) -> dict:
    """Return metadata dict from chdman info."""
    info = {}
    if not os.path.isfile(filepath):
        return info

    chdman = get_chdman_path()
    cmd = [chdman, "info", "-i", filepath]

    try:
        result = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            encoding="utf-8",
            errors="replace",
        )
        for line in result.stdout.splitlines():
            if ":" in line:
                key, _, val = line.partition(":")
                info[key.strip()] = val.strip()
    except Exception:
        pass
    return info

File: core/lib/src/validator.dart
// OpenROM — Universal ROM Compression Suite
// M5 Dev | GPL v3

import 'dart:convert';
import 'dart:io';
import 'detector.dart' as detector;
import 'logger.dart' as logger;

bool verifyChd(String chdPath, {void Function(String msg)? onLog}) {
  final logCb = onLog ?? logger.log;
  final file = File(chdPath);
  if (!file.existsSync()) {
    logCb('[ERROR] CHD file not found: $chdPath');
    return false;
  }

  final chdmanPath = detector.getChdmanPath();
  logCb('[VERIFY] Running chdman verify on $chdPath...');

  try {
    final result = Process.runSync(chdmanPath, ['verify', '-i', chdPath]);
    final output = '${result.stdout}\n${result.stderr}';
    final lines = const LineSplitter().convert(output);

    for (final line in lines) {
      if (line.trim().isNotEmpty) {
        logCb('  ${line.trim()}');
      }
    }

    if (result.exitCode == 0) {
      final lower = output.toLowerCase();
      if (lower.contains('is valid') ||
          lower.contains('integrity verified') ||
          lower.contains('raw sha1')) {
        return true;
      }
    }
    return result.exitCode == 0;
  } catch (e) {
    logCb('[ERROR] Failed to execute chdman verify: $e');
    return false;
  }
}
